File: problem4/vers_java/P4.java

public class P4
{
        public static void print( Object line )
        {
                System.out.print( line );
        }

        public static int check( int i, int j )
        {
                String prod;
                String front;
                String end;
                int half_length;
                int retval;

                prod = Integer.toString( i * j );
                half_length = prod.length() / 2;

                if( prod.length() % 2 == 0 ) {
                        front = prod.substring( 0, half_length );
                        end   = prod.substring( half_length, prod.length() );
                } else {
                        front = prod.substring( 0, half_length );
                        end   = prod.substring( half_length + 1, prod.length() );
                }

                end = new StringBuilder( end ).reverse().toString();

                if( front.equals( end ) )
                        retval = i * j;
                else
                        retval = -1;
                return retval;
        }

        public static void main( String args[] )
        {
                int i, j;
                int curr_pal;
                int largest_pal;
                int lowerbound, upperbound;

                if( args.length == 2 ) {
                        upperbound = Integer.parseInt( args[1] );
                        lowerbound = Integer.parseInt( args[0] );
                } else {
                        upperbound = 1000;
                        lowerbound = 100;
                }

                largest_pal = 0;
                for( i = lowerbound; i < upperbound; i++ ) {
                        for( j = i; j < upperbound; j++ ) {

                                curr_pal = check( i, j );
                                if( curr_pal > largest_pal )
                                        largest_pal = curr_pal;
                        }
                }

                print( "\nThe largest palindrome that is the sum " );
                print( "of two products between the bounds " + lowerbound );
                print( " and " + upperbound + " is: " + largest_pal + "\n\n" );
        }
}

File: problem4/vers_cpp/p4.cpp
/*
 * p4.cpp
 *
 * Zeke Reyna
 *
 * A simple C file to find the largest palindrome
 * That's a product of two 3 digit numbers.
 */

#include <iostream>
#include <string>
#include <algorithm>

#define SIZE 100
#define LOWERBOUND 100
#define UPPERBOUND 1000

using namespace std;

int check_if_pal(int i, int j)
{
	string num = to_string(i * j);
	string front;
	string end;
    int half_length;
    int retval;

    half_length = num.length() / 2;

	if (num.length() % 2 == 0) {
            front = num.substr(0, half_length);
            end   = num.substr(half_length, half_length);
    } else {
            front = num.substr(0, half_length);
            end   = num.substr(half_length + 1, half_length);
    }

    reverse(end.begin(), end.end());

    (front == end) ? (retval = i * j) : (retval = -1);

    return retval;
}

int main(int argc, char **argv)
{
	int i, j;
	int cur_pal;
	int largest_pal = 0;
	int lowerbound, upperbound;

	if (argc == 3) {
		upperbound = atoi(argv[2]);
		lowerbound = atoi(argv[1]);
	} else {
		upperbound = UPPERBOUND;
		lowerbound = LOWERBOUND;
	}

	for (i = lowerbound; i < upperbound; i++) {
		for (j = i; j < upperbound; j++) {

			cur_pal = check_if_pal(i, j);
			if (cur_pal > largest_pal)
				largest_pal = cur_pal;
		}
	}

	cout << "\nThe largest palindrome ";
	cout << "that is a sum of two products between ";
	cout << "the bounds " << lowerbound <<" and " << upperbound << " is: ";
	cout << largest_pal << "\n\n";

	return 0;
}
